diag_manager/fms_diag_object.F90

```diff
@@ -1,223 +1,115 @@
-     procedure :: has_units
-     procedure :: has_modname
-     procedure :: has_realm
-     procedure :: has_interp_method
-     procedure :: has_frequency
-     procedure :: has_output_units
-     procedure :: has_t
-     procedure :: has_tile_count
-     procedure :: has_area
-     procedure :: has_volume
-     procedure :: has_missing_value
-     procedure :: has_data_RANGE
-! Get functions
-     procedure :: get_diag_id => fms_diag_get_id
-     procedure :: get_attributes
-     procedure :: get_static
-     procedure :: get_registered
-     procedure :: get_mask_variant
-     procedure :: get_local
-     procedure :: get_vartype
-     procedure :: get_varname
-     procedure :: get_longname
-     procedure :: get_standname
-     procedure :: get_units
-     procedure :: get_modname
-     procedure :: get_realm
-     procedure :: get_interp_method
-     procedure :: get_frequency
-     procedure :: get_output_units
-     procedure :: get_t
-     procedure :: get_tile_count
-     procedure :: get_area
-     procedure :: get_volume
-     procedure :: get_missing_value
-     procedure :: get_data_RANGE
-!TODO     procedure :: get_init_time
-!TODO     procedure :: get_axis
+!TODO add container arrays
+private
+  class(fmsDiagFileContainer_type), allocatable :: FMS_diag_files (:) !< array of diag files
+  class(fmsDiagField_type), allocatable :: FMS_diag_fields(:) !< Array of diag fields
+  integer, private :: registered_variables !< Number of registered variables
+  logical, private :: initialized=.false. !< True if the fmsDiagObject is initialized
+  logical, private :: files_initialized=.false. !< True if the fmsDiagObject is initialized
+  logical, private :: fields_initialized=.false. !< True if the fmsDiagObject is initialized
+  logical, private :: buffers_initialized=.false. !< True if the fmsDiagObject is initialized
+  logical, private :: axes_initialized=.false. !< True if the fmsDiagObject is initialized
+
+  contains
+    procedure :: init => fms_diag_object_init
+    procedure :: fms_register_diag_field_scalar
+    procedure :: fms_register_diag_field_array
+    procedure :: fms_register_static_field
+    procedure :: register => fms_register_diag_field_obj !! Merely initialize fields.
+    procedure :: fms_diag_field_add_attribute
+    procedure :: fms_get_diag_field_id_from_name
+    procedure :: diag_end => fms_diag_object_end
 end type fmsDiagObject_type
-!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! variables !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
-type(fmsDiagObject_type) :: null_ob
-
-integer,private :: MAX_LEN_VARNAME
-integer,private :: MAX_LEN_META
-logical,private :: module_is_initialized = .false. !< Flag indicating if the module is initialized
 
-TYPE(fmsDiagObject_type), private, ALLOCATABLE, target :: diag_objs(:) !< Array of diag objects
-                                                                       !! one for each registered variable
+type (fmsDiagObject_type), target :: fms_diag_object
```

Comment:
I am wondering if should consider a convention for module singletons to get a prefix or suffix indicating they are such. This is because one can declare , as an exampe in this case, an fmsDiagObject_type and name it the logical/nice name fms_diag_object, and it would Technically be OK,  but it can lead to the user making a bug using a local object when the singleton was meant). Maybe call it fms_diag_object_ms? "_ms" for module singleton? . Maybe I'm worrying too much?